### ISSUE TITLE
Kernel: Scheduler "freaks out" when two processes frequently yield()

### ISSUE THREAD

**Issue by kleinesfilmroellchen** [SerenityOS/serenity#12679]
TL;DR: Scheduler appears to have problems when two processes at the same time call yield() a lot. CPU is at 100% but no application is actually doing much.

![image](https://user-images.githubusercontent.com/28656157/154838801-d5b5ff7e-5a8e-456d-8fc3-867d4accb7c7.png)

This profile looks ok, right? It's me starting and stopping a bunch of aplay audio playbacks, and those audio playbacks appear to not interfere with each other as they just use ~5-10% CPU, and that only in bursts. This is what Profiler sees after the fact in a system-wide profile. However, while the playback was happening, something really strange occurred: As soon as two aplay instances were active at the same time (and only as long as they were), the CPU would get pinned at 100%, which doesn't show up at all in the profile (not even AudioServer, who is doing very little all the time). I found out that both processes were yielding the CPU a bunch while waiting for the server to empty their playback queue. This is intentional design, or at least I couldn't think of some better way of telling the system "I can't do anything right now, please don't run this thread anymore". And with just one process doing it, everything seems fine, but as soon as two processes join (lol), I think the scheduler just gets overloaded with two processes that instantly yield the moment they get CPU time.

**Comment by ADKaster** [comment]
>  I couldn't think of some better way of telling the system "I can't do anything right now, please don't run this thread anymore".

The audio clients know about how much data they've sent to the server right? Sleeping for some amount of time or waiting to be woken up by a timer or blocking on a semaphore or some such would probably be better than repeatedly calling yield. Yield in that sense is essentially a busy loop in userspace, which is no bueno. If you want to block, then block 🤔.

**Comment by kleinesfilmroellchen** [comment]
The problem is that the yield happens in the shared queue, which doesn't know whether it's handling audio or whether it's dealing with data that is read at a regular rate. Additionally, the periods we're talking about are really brief, only 4-8ms. So having a timer for that is both overkill and imprecise. We can improve my new audio sending system after it's merged (please not before so that it actually gets merged in the first place) but the whole point about this queue is that it's lock-free and low-overhead, even for the blocking version. The blocking enqueue and dequeue calls can't have semaphores, because that would rely upon the assumption that the opposite operation was also blocking. Remember that the non-blocking operations assume real-time constraints, so no mutexes, no semaphores, no syscalls, just atomics. They can't signal the blocking operations to continue, at that point you could just have a normal blocking queue with two mutexes, that's not the point. The audio server uses the non-blocking dequeue while the client can use both blocking and non-blocking enqueue, the issue just arises with the blocking enqueue.

The only solution would then be to check the queue fullness outside the queue and wait there. This would still involve a spinlock of some variant as there's no one to signal us with a semaphore or condvar. So we're just moving the problem one level higher.

The Kernel is perfectly capable of handling one application with a pseudo-spinlock. Why shouldn't it be capable of handling two applications with pseudo-spinlocks?

**Comment by kleinesfilmroellchen** [comment]
Also, I think I didn't make the nature of "CPU is pinned at 100%" clear enough:

Because this is a yield() issue, the system doesn't become unresponsive. Other applications can increase in CPU usage and no difference is noticeable. Basically, the system behaves as if the yield() contention didn't exist, but 100% is reported and e.g. my processor fan does in fact start spinning up. It's very much a pseudo-load that doesn't hinder anything of importance.

**Comment by tomuta** [comment]
What @ADKaster said. You are actually wasting a lot of resources thi way, which is correctly reflected. The two threads keep bouncing each other, executing some code just to `yield` and kick it to the other one.

As @ADKaster also said, `yield` should be very, very, very rarely used (e.g. by the kernel itself when it blocks). You should be using other synchronization primitives. I am very sure what you're trying to achieve can be done without `yield`. A timer is indeed not the correct approach, but there is nothing wrong with using blocking primitives in audio processing (e.g. also does this, too).

So, bottom line, I am fairly sure the scheduler doesn't "freak out" here and indeed accurately reflects the waste of cpu resources induced by excessive calls to `yield`.
